godot_project/editor/rendering/particle_emitter_renderer/particle_emitter_renderer.gd

```diff
@@ -81,21 +81,53 @@ func transform_by_external_transform(in_selection_initial_pos: Vector3, in_initi
 	var delta_pos: Vector3 = in_initial_nano_struct_transform.origin - in_selection_initial_pos
 	var new_pos: Vector3 = in_external_transform.origin + in_external_transform.basis * delta_pos
 	global_transform = Transform3D(final_rotation.basis.orthonormalized(), new_pos)
-	_structure_preview.global_position = global_transform.origin
+	var emitter: NanoParticleEmitter = _workspace_context.workspace.get_structure_by_int_guid(_emitter_id)
+	for i: int in _structure_previews.size():
+		var preview: StructurePreview = _structure_previews[i]
+		preview.global_position = global_transform.origin + emitter.calculate_instance_offset(i)
+
+
+func _set_structure_preview_count(in_count: int) -> void:
+	const STRUCTURE_PREVIEW_SCN: PackedScene = preload("uid://dtf1gkl710hh8")
+	while in_count < _structure_previews.size():
+		var to_delete: StructurePreview = _structure_previews.pop_back()
+		to_delete.queue_free()
+	if in_count > _structure_previews.size():
+		var emitter: NanoParticleEmitter = _workspace_context.workspace.get_structure_by_int_guid(_emitter_id)
+		var template: AtomicStructure = emitter.get_parameters().get_molecule_template()
+		while in_count > _structure_previews.size():
+			var instance: StructurePreview = STRUCTURE_PREVIEW_SCN.instantiate() as StructurePreview
+			var index: int = _structure_previews.size()
+			instance.name = "Instance_%d" % index
+			instance.top_level = true
+			add_child(instance)
+			instance.set_structure(template)
+			instance.global_position = global_position + emitter.calculate_instance_offset(index)
+			instance.visible = self.visible
+			_structure_previews.push_back(instance)
 
 
 func _on_emitter_transform_changed(in_transform: Transform3D) -> void:
 	global_transform = in_transform
-	_structure_preview.global_position = in_transform.origin
+	var emitter: NanoParticleEmitter = _workspace_context.workspace.get_structure_by_int_guid(_emitter_id)
+	for i: int in _structure_previews.size():
+		var preview: StructurePreview = _structure_previews[i]
+		preview.global_position = in_transform.origin + emitter.calculate_instance_offset(i)
```

Comment:
Suggestion: Since `calculate_instance_offset` will always return the same result, it could be cached in each `StructurePreview` directly, and updated when the preview count changes. This way there's no need to redo all the expensive calculations each frame while the emitter is moving.